### ISSUE TITLE
Multipart MIME parser may swallow MIME errors depending on stream chunk size

### ISSUE THREAD

**Issue by RichardBradley** [akka/akka#17051]
I have found that the `akka.http.unmarshalling.MultipartUnmarshallers` will behave quite differently with invalid MIME streams depending on how the stream has been divided into chunks.

I have uploaded a failing test to https://gist.github.com/RichardBradley/0e73adaa86c4d377bf65/95ff8af7298ad64808700fee59e736a4c1e10031

(Run it with `sbt test`)

I have picked out one example: when a MIME body part is missing its headers, the MIME parser will throw a fairly helpful error if the request is `Strict`:
```
akka.http.model.ParsingException: Illegal character ' ' in header name
```

However, if the same MIME stream is `Default` or `Chunked`, then the error will not be thrown. Instead the BodyPart stream will be truncated (without error) at varying lengths depending on the exact chunk size used.

**Comment by RichardBradley** [comment]
@sirthias -- did you manage to get anywhere with this? I tried to look but couldn't figure out where the bug might lie.

**Comment by jrudolph** [comment]
@RichardBradley see PR #17121.

**Comment by RichardBradley** [comment]
Thanks, sorry for not spotting that.

Do you know why that PR hasn't been merged and this issue closed? Is it just in a queue?

**Comment by jrudolph** [comment]
It's just still in the queue, as you can see the list of open PRs is quite long and the easter holidays took their toll...

**Comment by RichardBradley** [comment]
OK, thanks very much.
Sorry for the noise. :-)

**Comment by RichardBradley** [comment]
The test still fails for me on Akka `1.0-RC1`.

The behaviour no longer varies by the streaming chunk size, but the MIME message is truncated on error, whereas I would expect the stream to fail with a MIME parsing error.

When the entity is Strict, unmarshalling does fail with a MIME parsing error. I think it is a bug for streamed entities to swallow these errors.

I have updated the test case at https://gist.github.com/RichardBradley/0e73adaa86c4d377bf65 -- please could you retest and reopen this issue if you agree?

**Comment by jrudolph** [comment]
This seems to depend on the preamble, that's also why the existing tests didn't catch it.

**Comment by jrudolph** [comment]
There's more broken that this. Another test that fails is this one:

```scala
      "an implicitly typed part (without headers) (Default)" in {
        val content = """--XYZABC
                        |
                        |Perfectly fine part content.
                        |--XYZABC--""".stripMarginWithNewline("\r\n")
        val byteStrings = content.map(c ⇒ ByteString(c.toString)) // one-char ByteStrings
        Unmarshal(HttpEntity.Default(`multipart/mixed` withBoundary "XYZABC", content.length, Source(byteStrings)))
          .to[Multipart.General] should haveParts(
            Multipart.General.BodyPart.Strict(HttpEntity(ContentTypes.`text/plain(UTF-8)`, "Perfectly fine part content.")))
      }
```

which is just one failing example where the first boundary itself is chunked. It seems in `tryParseInitialBoundary` too much data is thrown away in the continue call.

**Comment by jrudolph** [comment]
I pushed a partial fix that already fixes the new issue but not the one by @RichardBradley

**Comment by jrudolph** [comment]
PR #17654

**Comment by jrudolph** [comment]
Closed by #17654.

**Comment by RichardBradley** [comment]
Thanks!
